### ISSUE TITLE
Can't find bundler/shared_helpers after upgrade to jruby 9.2.0.0

### ISSUE THREAD

**Issue by rsov** [jruby/warbler#432]
Hello

This issue occured after upgrading to new version of jruby

Stack:
jruby 9.2.0.0
rails  4.2.10
java 1.8
tomcat 8.5.31
warbler  2.0.5

War file builds fine but when deployed to tomcat it throws an error: 

```
org.jruby.rack.RackInitializationException: no such file to load -- bundler/shared_helpers
	from org/jruby/RubyKernel.java:970:in `require'
	from uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:59:in `require'
	from file:/usr/local/tomcat-data/webapps/secure/META-INF/init.rb:25:in `<eval>'
	from org/jruby/RubyKernel.java:1037:in `eval'
	from uri:classloader:/jruby/rack/booter.rb:256:in `block in load_settings_from_init_rb'
	from org/jruby/RubyArray.java:1801:in `each'
	from uri:classloader:/jruby/rack/booter.rb:243:in `load_settings_from_init_rb'
	from uri:classloader:/jruby/rack/booter.rb:99:in `boot!'
	from uri:classloader:/jruby/rack/rails_booter.rb:26:in `boot!'
	from uri:classloader:/jruby/rack/boot/rails.rb:10:in `<main>'
	from org/jruby/RubyKernel.java:994:in `load'
	from <script>:1:in `<main>'

	at org.jruby.rack.RackInitializationException.wrap(RackInitializationException.java:29)
	at org.jruby.rack.RackApplicationFactoryDecorator.init(RackApplicationFactoryDecorator.java:104)
	at org.jruby.rack.RackServletContextListener.contextInitialized(RackServletContextListener.java:50)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4792)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5256)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:754)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:730)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:734)
	at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1140)
	at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1875)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.jruby.exceptions.LoadError: (LoadError) no such file to load -- bundler/shared_helpers
	at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:970)
	at RUBY.require(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:59)
	at RUBY.<eval>(file:/usr/local/tomcat-data/webapps/secure/META-INF/init.rb:25)
	at org.jruby.RubyKernel.eval(org/jruby/RubyKernel.java:1037)
	at RUBY.block in load_settings_from_init_rb(uri:classloader:/jruby/rack/booter.rb:256)
	at org.jruby.RubyArray.each(org/jruby/RubyArray.java:1801)
	at RUBY.load_settings_from_init_rb(uri:classloader:/jruby/rack/booter.rb:243)
	at RUBY.boot!(uri:classloader:/jruby/rack/booter.rb:99)
	at RUBY.boot!(uri:classloader:/jruby/rack/rails_booter.rb:26)
	at RUBY.<main>(uri:classloader:/jruby/rack/boot/rails.rb:10)
	at org.jruby.RubyKernel.load(org/jruby/RubyKernel.java:994)
	at RUBY.<main>(<script>:1)
```

To my understanding warbler should load the jruby environment prior to calling this script and have all the paths initialized. 

I tried using `load` and specifying path directly to that file, which loaded it fine but then threw a different `require` error from within loaded file

Another thing I've noticed: The RubyGems does not have the corect Gem.path. 
Gem.path are:
```
["/usr/share/tomcat6/.gem/jruby/2.5.0", "uri:classloader:/META-INF/jruby.home/lib/ruby/gems/shared"]
```
Should be:
```
["/usr/local/tomcat-data/webapps/secure/WEB-INF/gems", "uri:classloader:/META-INF/jruby.home/lib/ruby/gems/shared"]
```

**Comment by rsov** [comment]
So I found a "fix"

Putting a few lines in the `META-INF/init.rb`

After all of the logic is done and before the `require 'bundle/shared_helpers'` add these lines:

```
Gem.clear_paths
Gem.path
```

This reloads the paths and sets them again from the above set `ENV`

**Comment by kares** [comment]
expected to be resolved with https://github.com/jruby/warbler/pull/433, also the issues looks similar to an older [one](https://github.com/jruby/warbler/pull/430) but that had a different resolution at some point ...

**Comment by kares** [comment]
^^^ already reverted the proposed fix that used seemingly non-existent RubyGems method ...

while the thinking here might be fine, it might break some embedded cases where unpacking does not happen. there should be no need to fiddle around `Gem.path` in the initialization script - proper fix ends up at a higher level. while work-arounds are welcome, starting to think this is a JRuby regression that might get fixed in 9.2.1 ... we need to understand why it regressed (maybe related to change in __FILE__ canicalization?)

**Comment by rsov** [comment]
Sorry guys. It is my bad. I was really excited to push up the fix without giving a proper go over. (even thought it is just one line of code)

**Comment by kares** [comment]
I do not see which Bundler version was used? If it is wasn't **1.16.3** please try it out and report back ...

**Comment by rsov** [comment]
Tried it with 1.16.1 and 1.16.2 - both consistent. I am not sure where to get 1.16.3

Putting: `gem  'bundler', git: 'https://github.com/bundler/bundler'` in Gemfile leads to following error:

```
[!] There was an error while loading `bundler.gemspec`: Could not locate Gemfile. Bundler cannot continue.

 #  from /home/ruslan/.rvm/gems/jruby-9.2.0.0@tservices/bundler/gems/bundler-8d53c82ab384/bundler.gemspec:1
 #  -------------------------------------------
 >  # coding: utf-8
 #  # frozen_string_literal: true
 #  -------------------------------------------
```

**Comment by kares** [comment]
sorry, did not realize it isn't yet released ... this might resolve issues https://github.com/bundler/bundler/pull/6502
`__FILE__` canonicalization changed in 9.2 - so if that would work we wouldn't need to do much here.

**Comment by rsov** [comment]
ah perfect! Will test with it when it comes out

**Comment by EugenMayer** [comment]
really waiting for 9.2 compat, great your a on it @kares !

**Comment by timbogit** [comment]
@kares I tried using bundler 1.16.3 (which contains the fix for https://github.com/bundler/bundler/pull/6502, for all I can tell), and the warbler output still appears to show the same issue when used with JRuby 9.2.0.0 (i.e., `org.jruby.rack.RackInitializationException: no such file to load -- bundler/shared_helpers`).

Given that JRuby 9.2.1.0 seems a little was out, do you have any suggestions around how to work around this issue?

**Comment by dennis-d** [comment]
#437 seems to fix the issue

**Comment by lauborges** [comment]
Based on @dennis-d's tip
I have added the line
`config.init_contents << StringIO.new("\nGem.clear_paths\nGem.path\n\n")` 
to warbler.rb and it works

**Comment by andy-twosticks** [comment]
The warble.rb fix does not work for me.  When can we expect this #437 fix to appear in a Gemfile so that I can use it in production?

(Bundler 1.16.2, warbler 2.0.5.)

Detail:

@dennis-d 's change just gets me a syntax error when I run warble.  @lauborges change allows me to run warble successfully, but the resultant .war file just gives an internal server error:


    Cannot find a config file from /var/lib/tomcat/webapps/warbler_test/WEB-INF/core/lib/boot.rb:174:in `get_config' from /var/lib/tomcat/webapps/warbler_test/WEB-INF/core/init.rb:33:in `<module:SwingShift>' from /var/lib/tomcat/webapps/warbler_test/WEB-INF/core/init.rb:30:in `<main>' from org/jruby/RubyKernel.java:970:in `require' from /var/lib/tomcat/webapps/warbler_test/WEB-INF/app.rb:1:in `<main>' from org/jruby/RubyKernel.java:970:in `require' from /var/lib/tomcat/webapps/warbler_test/WEB-INF/app.rb:3:in `block in (root)' from org/jruby/RubyBasicObject.java:2605:in `instance_eval' from /usr/share/tomcat/webapps/warbler_test/WEB-INF/config.ru:1:in `(root)' from /usr/share/tomcat/webapps/warbler_test/WEB-INF/config.ru:9:in `<main>' 

Here's my warble.rb:

    Warbler::Config.new do |config|
      config.init_contents << StringIO.new("\nGem.clear_paths\nGem.path\n\n")
      config.dirs = %w(config core projects public db)
      config.includes = FileList[ 'config.ru',
                                  'controllers.rb',
                                  'version.rb',
                                  'app.rb' ]

      config.features = %w(runnable)
      config.webxml.booter = :rack
    end

**Comment by mungler** [comment]
@lauborges fix works for me with JRuby 9.2.1.0 - thanks!

**Comment by deivid-rodriguez** [comment]
I think this was fixed in jruby 9.2.11.0 by https://github.com/jruby/jruby/pull/6084, since [the error I reported there](https://github.com/jruby/jruby/issues/6083) was exactly the same.

So I think this can be closed.

In any case, I don't think the bundler monkeypatch causing this error is needed anymore so it might get removed and the issue gone for all jrubies.

**Comment by olleolleolle** [comment]
@deivid-rodriguez Thanks for the detailed report & notes. I will close this, with that explanation. 💚
